### PR TITLE
enable tensor parallelism for MXLinear

### PR DESCRIPTION
**Pull request by vkuzo**
Summary:

Enables TP for MXLinear. Specifically:
1. change the reshape logic from `x.reshape(-1, block_size)` to
   `x.reshape(*orig_shape[:-1], orig_shape[-1] // block_size,
   block_size)`
2. modify the rest of the code to adhere to (1)
3. cast input tensor and max_abs to float32 before calculating the MX
   scale, in order to get around another bug in DTensor + view + int16
   target type

(1) is necessary because the old reshape logic would flatten dims, which
did not work if one of those flattened dims was sharded.

Note that TP does not yet work with the custom dim1 triton kernel, we'll need a separate PR to fix that by adding a sharding strategy to the kernel.

I verified that performance for FSDP + mxfp8 + compile is not affected by this stack, with torchtitan llama 3 8B on 8 B200 GPUs:

```bash
baseline (without this PR stack)

bf16 FSDP - tps 8.8k, peak_mem 35.0 GiB ([link](https://www.internalfb.com/phabricator/paste/view/P1850041288))
with-proxy CONFIG_FILE="torchtitan/models/llama3/train_configs/llama3_8b.toml " ./run_train.sh --model.print_after_conversion --training.steps 50 --parallelism.tensor_parallel_degree=1 --training.compile

bf16 FSDP + tp - tps 8.2k, peak_mem 29.6 GiB ([link](https://www.internalfb.com/phabricator/paste/view/P1850041882))
with-proxy CONFIG_FILE="torchtitan/models/llama3/train_configs/llama3_8b.toml " ./run_train.sh --model.print_after_conversion --training.steps 50 --parallelism.tensor_parallel_degree=2 --training.compile

mxfp8 FSDP - tps 10k, peak_mem 35.3 GiB ([link](https://www.internalfb.com/phabricator/paste/view/P1850040695))
with-proxy CONFIG_FILE="torchtitan/models/llama3/train_configs/llama3_8b.toml " ./run_train.sh --model.print_after_conversion --training.steps 50 --model.converters mx --mx.recipe_name "mxfp8" --parallelism.tensor_parallel_degree=1 --mx.use_fp8_dim1_cast_triton_kernel --training.compile

mxfp8 FSDP + TP - broken
with-proxy CONFIG_FILE="torchtitan/models/llama3/train_configs/llama3_8b.toml " ./run_train.sh --model.print_after_conversion --training.steps 50 --model.converters mx --mx.recipe_name "mxfp8" --parallelism.tensor_parallel_degree=2 --mx.use_fp8_dim1_cast_triton_kernel --training.compile

experiment (with this PR stack)

mxfp8 FSDP - tps 10k, peak_mem 35.3 GiB ([link](https://www.internalfb.com/phabricator/paste/view/P1850044437))
with-proxy CONFIG_FILE="torchtitan/models/llama3/train_configs/llama3_8b.toml " ./run_train.sh --model.print_after_conversion --training.steps 50 --model.converters mx --mx.recipe_name "mxfp8" --parallelism.tensor_parallel_degree=1 --mx.use_fp8_dim1_cast_triton_kernel --training.compile

mxpf8 + FSDP + TP + turn off dim triton kernel - tps 7.9k, peak_mem 29.7 GiB ([link](https://www.internalfb.com/phabricator/paste/view/P1850045992))
with-proxy CONFIG_FILE="torchtitan/models/llama3/train_configs/llama3_8b.toml " ./run_train.sh --model.print_after_conversion --training.steps 50 --model.converters mx --mx.recipe_name "mxfp8" --parallelism.tensor_parallel_degree=2 --mx.no-use_fp8_dim1_cast_triton_kernel --training.compile
```

Test Plan:

```
pytest test/prototype/mx_formats
./test/prototype/mx_formats/test_dtensor.sh
```

Reviewers:

Subscribers:

Tasks:

Tags:

### CODE REVIEW

**Comment by vkuzo** [comment]
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #2436
* __->__ #2434

File: test/prototype/mx_formats/test_mx_linear.py [hunk 1]
```diff
@@ -190,8 +190,8 @@ def test_linear_eager_emulated_vs_real_gemm(recipe_name, mkn):
 # TODO(future): enable compile support
 @pytest.mark.skipif(not torch.cuda.is_available(), reason="CUDA not available")
 def test_activation_checkpointing():
-    input_shape = (2, 4)
-    grad_shape = (2, 8)
+    input_shape = (16, 4)
```
Comment (vkuzo) [inline test/prototype/mx_formats/test_mx_linear.py:193]: this was broken before, caught by enforcing that inner dim is divisible by block size

File: torchao/prototype/mx_formats/mx_tensor.py [hunk 2]
```diff
@@ -160,22 +156,33 @@ def to_mx(
...
+
+    # We cast to float32 here because
+    # in the `max_abs_int32 = max_abs.view(hp_int_dtype)` line below,
+    # if tensor parallel is enabled then the resulting shape is 2x larger
+    # than it should be under some conditions, likely because of a bug in
+    # the `view` op with DTensor and target dtype int16.  I reproduce in
+    # torchtitan but not in a unit test, so not enough info to file a good
+    # issue in pytorch/pytorch. For now, work around. In the future we should
+    # debug and fix this properly.
+    data_hp = data_hp.to(torch.float32)
```
Comment (vkuzo) [inline torchao/prototype/mx_formats/mx_tensor.py:184]: performance testing showed that with compile on, having this in float32 does not regress performance

File: torchao/testing/training/dtensor_utils.py [hunk 3]
```diff
@@ -152,15 +152,18 @@ def _test_lowp_mlp_tensor_parallelism_base(
...
     x_fp32 = torch.rand(size, size * 2, size, device=device, requires_grad=False)
+    go_fp32 = torch.rand(size, size * 2, size, device=device, requires_grad=False)
     x_fp32_tp_input = x_fp32.clone()
+    go_fp32_tp = go_fp32.clone()
     x_fp32_sp_input = distribute_tensor(x_fp32.clone(), mesh, [Shard(0)])
+    go_fp32_sp = distribute_tensor(go_fp32.clone(), mesh, [Shard(0)])
 
     tp_out = tp_model(x_fp32_tp_input)
-    tp_out.sum().backward()
+    tp_out.backward(go_fp32_tp)
```
Comment (vkuzo) [inline torchao/testing/training/dtensor_utils.py:162]: to make sure grad flowing into the last linear is contiguous
